README
NAME
    Package::Strictures - Facilitate toggling validation code at users
    request, without extra performance penalties.

VERSION
    version 1.000000

SYNOPSIS
  IMPLEMENTING MODULES
      package Foo::Bar::Baz;

      use Package::Strictures::Register -setup => {
          -strictures => {
              STRICT => {
                default => ''
              },
          },
      };

      if( STRICT ) {
        /* Elimintated Code */
      }

    See "Package::Strictures::Register" for more detail.

  CONSUMING USERS
      use Package::Strictures -for => {
        'Foo::Bar::Baz' => {
          'STRICT' => 1,
        },
      };

      use Foo::Bar::Baz;

      /* Previously eliminated code now runs.

DESCRIPTION
    Often, I find myself in a bind, where I have code I want to do things
    properly, so it will detect of its own accord ( at run time ) misuses of
    varying data-structures or methods, but the very same tools that would
    be used to analyze and assure that things are going correctly, result in
    substantial performance penalties.

    This module, and the infrastructure I hope builds on top of it, may
    hopefully provide an 'in' that lets me have the best of both worlds,
    fast on the production server, and concise when trying to debug it (
    that is, not having to manually desk-check the whole execution cycle
    through various functions and modules just to find which level things
    are going wrong at ).

    In an ideal world, code would be both fast and concise, however, that is
    a future fantasy, and this here instead aims to produce 80% of the same
    benefits, but now, instead of never.

MINOR WARNING
    This code is pretty fresh. Its been done to death with T.D.D., and does
    everything I want it to, but there's always bugs.

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Package-Strictures

1.000002 2017-03-10T14:24:09Z 0a9de4d

1.000001 2017-03-10T14:14:38Z 0c6ec84
 - Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: +1 ↓1 (recommends: ↑1)
 - develop: +18 ↑4 -6 (recommends: +1, suggests: +2 -1)
 - runtime: ↓1 -1
 - test: +2 ↓1 -3 (recommends: ↑2 ↓1 -1)

1.000000 2014-04-30T12:36:53Z
 [00 Major]
 - A few moderate implementation/dependency changes
 - Packaging changes that might be important for downstreams repackaging by hand.

 [01 New Constant generation bug fix]
 - Instead of the previous approach of trying to create constants for the per-package stricutres ourselves, we now use
   a combination of constant.pm and Import::Into
 - This approcach should be more reliable long term for ensuring constant folding takes place.
 - This change resolves closes #1 on github.
 - This also means this package is no longer broken >perl 5.19.3 after the close of perl #79908
 - Thanks to Slaven Rezić for reporting, and thanks to Andreas J. Koenig for the scattered smoke testing that helped me
   find which Perl change broke this.

 [02 Moose requirement eliminated]
 - Moose was previously used in Registry.pm to do almost nothing, and so the dependency was easy to eliminate.
 - namespace::autoclean was nuked in favour of namespace::clean to further eliminate Moose from installation
   requirements.

 [03 Tooling switched]
 - ExtUtils::MakeMaker is now used instead of Module::Build
 - And "Latest" versions are now only recommended dependencies instead of mandatory ones.

 [04 Version Scheme switched]
 - Version scheme reduced to floating point with a mantissa of 6, as opposed to a mantissa of 8
 - This is the primary reason for 0.xxx -> 1.xxx

 [Dependencies::Aggregate]
 - 5.008 : 48
 - 5.010 : 37
 - 5.018 : 22
 - 5.019 : 23

 [Dependencies::Stats]
 - Dependencies changed since 0.01001319, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +44 -2 (recommends: -1, suggests: +1 -1)
 - runtime: +3 ↑1 -3
 - test: +6 ↓1 -2 (recommends: +4)

0.01001319 2012-12-02T14:23:20Z
 - Maintenance release.

 [Dependencies::Stats]
 - Dependencies changed since 0.01001318, see misc/*.deps* for details
 - build: +1
 - configure: +1
 - develop: +6 (recommends: +1, suggests: +1)
 - runtime: +3 -4
 - test: +6

 [Documentation]
 - regenerated, which removes that annoying 'our $VERSION' from the examples =p
 - DESCRIPTION moved to after SYNOPSIS
 - Copyright years updated.

 [Internals]
 - $VERSION declaration syntax updated
 - All packages declare $AUTHORITY

 [Packaging]
 - bundle .perltidyrc
 - move dist.ini to bundle-based config.
 - Changes moved to CPAN::Changes format.
 - Updated LICENSE copyright year.
 - Updated LICENSE indentation.
 - Updated LICENSE address.
 - Moved author/release tests from t/ and put in xt/
 - update git metadata
 - set bug tracker to github

 [Tests]
 - Removed Test::Exception based code in favour of Test::Fatal

 [Tests::Moved]
 - t/author-critic.t -> xt/author/critic.t
 - t/author-kwalitee -> xt/release/kwalitee.t
 - t/release-meta-yaml.t -> xt/release/distmeta.t
 - t/release-pod-coverage.t -> xt/release/pod-coverage.t
 - t/release-pod-syntax.t -> xt/release/pod-syntax.t

 [Tests::New]
 - @install : 0000-report-versions-tiny.t
 - @release : cpan-changes.t
 - @release : eol.t
 - @release : minimum-version.t

0.01001318 2009-11-10T22:49:10Z
 - Backed NS::AC from 0.9 to 0.09 , that always gets me.
 - Nuked Strange Unicode in POD.
 - Dropped MakeMaker.pl and just use MB

 [Dependencies::Stats]
 - Dependencies changed since 0.01000019, see misc/*.deps* for details
 - runtime: ↓1

0.01000019 2009-10-28T23:22:41Z
 - Fixed Dependencies.

 [Dependencies::Stats]
 - Dependencies changed since 0.01000007, see misc/*.deps* for details
 - runtime: +1

0.01000007 2009-10-28T11:49:12Z
 - First version.


t/01-poc-lib/Example.pm
package    # NO CPAN
  Example;

use strict;
use warnings;

use Package::Strictures::Register -setup => {
  -strictures => {
    'STRICT' => {
      default => '',
      type    => 'Bool',              # NOT IMPLEMENTED OR TESTED
      'ENV'   => 'EXAMPLE_STRICT',    # NOT IMPLEMENTED OR TESTED
    }
  },
  -groups => {                        # NOT IMPLEMENTED OR TESTED
    '@all' => {
      doinstead => { STRICT => 1, },
      ENV       => 'EXAMPLE_STRICT_ALL',
    }
  }
};

use namespace::autoclean;

sub slow {
  my $result = 5;

  if (STRICT) {

    # Simulate some slow validation.
    #
    sleep 1;

    if (@_) {
      die "slow() takes no parameters";
    }
  }
  return $result;
}

1;



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "B::Deparse" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.4003",
    "Test::Fatal" => 0,
    "Test::More" => "0.98"
  },
  "configure_requires" => {
    "Module::Build" => "0.4003"
  },
  "dist_abstract" => "Facilitate toggling validation code at users request, without extra performance penalties.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Package-Strictures",
  "dist_version" => "0.01001319",
  "license" => "perl",
  "module_name" => "Package::Strictures",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Config::INI::Reader" => "0.014",
    "Moose" => "0.92",
    "MooseX::ClassAttribute" => "0.10",
    "namespace::autoclean" => "0.09",
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Package/Strictures.pm
use 5.006;    #6 = pragmas, our, 5 = qr, 4 = __PACAKGE__,for my
use strict;
use warnings;

package Package::Strictures;

our $VERSION = '1.000002';

# ABSTRACT: Facilitate toggling validation code at users request, without extra performance penalties.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Package::Strictures::Registry;
use Carp ();























































sub import {
  my ( $self, %params ) = @_;
  if ( not %params ) {
    Carp::carp( __PACKAGE__ . ' called with no parameters, skipping magic' );
    return;
  }
  if ( ( not exists $params{-for} ) && ( not exists $params{-from} ) ) {
    Carp::croak( 'no -for or -from parameter to ' . __PACKAGE__ );
  }
  if ( exists $params{-for} ) {
    $self->_setup_for( $params{-for} );
    return;
  }
  if ( exists $params{-from} ) {
    $self->_setup_from( $params{-from} );
    return;
  }
  Carp::croak('Ugh!?');
}

sub _setup_for {
  my ( $self, $params ) = @_;
  my $reftype = ref $params;
  if ( $reftype ne 'HASH' ) {
    Carp::croak("-for presently only takes HASH, got `$reftype`");
  }
  for my $package ( keys %{$params} ) {
    $self->_setup_for_package( $params->{$package}, $package );
  }
  return;
}

sub _setup_from {
  my ( $self, $from ) = @_;
  my $reftype = ref $from;
  if ($reftype) {
    Carp::croak("-from can only take a scalar, not a ref, got `$reftype`");
  }
  ## no critic (RegularExpressions::RequireDotMatchAnything RegularExpressions::RequireLineBoundaryMatching)
  if ( $from =~ /[.]ini$/x ) {
    $self->_setup_from_ini($from);
    return;
  }
  Carp::croak('Only know how to load setup from .ini files at present');
  return;
}

sub _setup_from_ini {
  my ( $self, $from ) = @_;
  require Config::INI::Reader;
  my ($conf) = Config::INI::Reader->read_file($from);
  if ( exists $conf->{'_'} ) {
    Carp::carp("Strictures not addressed at a package found in `$from`");
  }
  $self->_setup_for($conf);
  return;
}

sub _setup_for_package {
  my ( undef, $params, $package ) = @_;
  my $reftype = ref $params;
  if ( $reftype ne 'HASH' ) {
    my $mesg = q[];
    $mesg .= q[-for => { Some::Name => X } presently only takes HASH,];
    $mesg .= q[got `%s` on package `%s` ];
    Carp::croak( sprintf $mesg, $reftype, $package );
  }
  for my $value ( keys %{$params} ) {
    Package::Strictures::Registry->set_value( $package, $value, $params->{$value} );
  }
  return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Package::Strictures - Facilitate toggling validation code at users request, without extra performance penalties.

=head1 VERSION

version 1.000002

=head1 SYNOPSIS

=head2 IMPLEMENTING MODULES

  package Foo::Bar::Baz;

  use Package::Strictures::Register -setup => {
      -strictures => {
          STRICT => {
            default => ''
          },
      },
  };

  if( STRICT ) {
    /* Elimintated Code */
  }

See L<< C<Package::Strictures::Register>|Package::Strictures::Register >> for more detail.

=head2 CONSUMING USERS

  use Package::Strictures -for => {
    'Foo::Bar::Baz' => {
      'STRICT' => 1,
    },
  };

  use Foo::Bar::Baz;

  /* Previously eliminated code now runs.

=head1 DESCRIPTION

Often, I find myself in a bind, where I have code I want to do things properly, so it will detect
of its own accord ( at run time ) misuses of varying data-structures or methods, but the very same
tools that would be used to analyze and assure that things are going correctly, result in substantial
performance penalties.

This module, and the infrastructure I hope builds on top of it, may hopefully provide an 'in' that lets me have the best of both
worlds, fast on the production server, and concise when trying to debug it ( that is, not having to manually desk-check the
whole execution cycle through various functions and modules just to find which level things are going wrong at ).

In an ideal world, code would be both fast and concise, however, that is a future fantasy, and this here instead aims to
produce 80% of the same benefits, but now, instead of never.

=head1 MINOR WARNING

This code is pretty fresh. Its been done to death with T.D.D., and does everything I want it to, but there's always bugs.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Package/Strictures/Register.pm
use 5.006;    # 6 = pragmas, our, 4 = __PACAKGE__
use strict;
use warnings;

package Package::Strictures::Register;

our $VERSION = '1.000002';

# ABSTRACT: Create compile-time constants that can be tweaked by users with Package::Strictures.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Package::Strictures::Registry ();
use Carp                          ();

sub import {
  my ( $self, %params ) = @_;

  if ( not %params ) {
    Carp::carp( __PACKAGE__ . ' called with no parameters, skipping magic' );
    return;
  }

  my (@caller) = caller;
  my $package = $caller[0];
  $package = $params{-for} if exists $params{-for};

  if ( not exists $params{-setup} ) {
    Carp::croak("Can't setup strictures for package '$package', need -setup ");
  }

  $self->_setup( $params{-setup}, $package );

  return;
}

sub _setup {
  my ( $self, $params, $package ) = @_;

  my $reftype = ref $params;

  if ( not 'HASH' eq $reftype ) {
    Carp::croak(qq/ -setup => can presently only support a HASH. Got '$reftype'/);
  }

  if ( ( not exists $params->{-strictures} ) && ( not exists $params->{-groups} ) ) {
    Carp::croak('Neither -setup => { -strictures }  or -setup => { -groups } provided.');
  }

  if ( exists $params->{-groups} ) {
    Carp::carp('-groups support is not yet implemented');
  }

  $params->{-strictures} = {} unless exists $params->{-strictures};
  $params->{-groups}     = {} unless exists $params->{-groups};

  $self->_setup_strictures( $params->{-strictures}, $package );

  return;
}

sub _setup_strictures {
  my ( $self, $strictures, $package ) = @_;
  my $reftype = ref $strictures;

  if ( not 'HASH' eq $reftype ) {
    Carp::croak( qq/Can't handle anything except a HASH ( Got $reftype )/
        . qq/ for param -setup => { -strictures =>  } in -setup for $package/ );
  }

  for my $subname ( keys %{$strictures} ) {

    $self->_setup_stricture( $strictures->{$subname}, $package, $subname );
  }

  return;
}

sub _setup_stricture {
  ## no critic 'ProhibitNoStrict'
  my ( $self, $prototype, $package, $name ) = @_;
  if ( not exists $prototype->{default} ) {
    Carp::croak("Prototype for `$package`::`$name` lacks a [required] ->{'default'} ");
  }

  $self->_advertise_stricture( $package, $name );

  require Import::Into;
  require constant;

  constant->import::into( $package, $name, $self->_fetch_stricture_value( $package, $name, $prototype->{default} ) );

  return;
}

sub _advertise_stricture {
  my ( undef, $package, $name ) = @_;
  Package::Strictures::Registry->advertise_value( $package, $name );
  return;
}

sub _fetch_stricture_value {
  my ( undef, $package, $name, $default ) = @_;
  if ( Package::Strictures::Registry->has_value( $package, $name ) ) {
    return Package::Strictures::Registry->get_value( $package, $name );
  }
  return $default;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Package::Strictures::Register - Create compile-time constants that can be tweaked by users with Package::Strictures.

=head1 VERSION

version 1.000002

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Package/Strictures/Registry.pm
use 5.006;    #  6 = pragmas, our, 5 = list undef.
use strict;
use warnings;

package Package::Strictures::Registry;

our $VERSION = '1.000002';

# ABSTRACT: Data Storage name-space for stricture parameters.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Carp ();

my $_registry_store = {};

sub _has_package {
  my ( undef, $package ) = @_;
  return exists $_registry_store->{$package};
}

sub _set_package {
  my ( undef, $package, $value ) = @_;
  return $_registry_store->{$package} = $value;
}

sub _get_package {
  my ( undef, $package ) = @_;
  return $_registry_store->{$package};
}

my $_advertisements = {};

sub _has_advert {
  my ( undef, $package ) = @_;
  return exists $_advertisements->{$package};
}

sub _set_advert {
  my ( undef, $package, $value ) = @_;
  return $_advertisements->{$package} = $value;
}

sub _get_advert {
  my ( undef, $package ) = @_;
  return $_advertisements->{$package};
}














sub advertise_value {
  my ( $self, $package, $name ) = @_;
  if ( not $self->_has_advert($package) ) {
    $self->_set_advert( $package, {} );
  }
  if ( not exists $self->_get_advert($package)->{$name} ) {
    $self->_get_advert($package)->{$name} = \$name;
  }
  else {
    Carp::croak("` $package :: $name` is already advertised!");
  }
  return;
}











sub has_value {
  my ( $self, $package, $name ) = @_;
  return unless ( $self->_has_package($package) );
  return exists $self->_get_package($package)->{$name};
}












sub get_value {
  my ( $self, $package, $name ) = @_;
  if ( not $self->has_value( $package, $name ) ) {
    Carp::croak("Error: package `$package` is not in the registry");
  }
  return $self->_get_package($package)->{$name};
}











sub set_value {
  my ( $self, $package, $name, $value ) = @_;
  if ( not $self->_has_package($package) ) {
    $self->_set_package( $package, {} );
  }
  $self->_get_package($package)->{$name} = $value;
  return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Package::Strictures::Registry - Data Storage name-space for stricture parameters.

=head1 VERSION

version 1.000002

=head1 METHODS

=head2 advertise_value

  Package::Strictures::Registry->advertise_value( 'Some::Package', "STRICT");

An informational data-storage for developers to see what packages that are loaded have strictures that are able to be tuned,
without having to grok the source.

Note that by the time you see this value, it is already too late to try setting it.

=head2 has_value

  Package::Strictures::Registry->has_value( 'Some::Package', 'STRICT' )

Sees if somebody ( A developer ) has defined an override value for the stricture.

This will be picked up by a performing package when somebody first use/require's it.

=head2 get_value

  Package::Strictures::Registry->get_value('Some::Package' , 'STRICT' )

Returns the value stored earlier if there was one.

This is done internally by L<<  C<Package::Strictures::Register>|Package::Strictures::Register >> to populate the values for
the compile-time constants.

=head2 set_value

  Package::Strictures::Registry->set_value('Some::Package', 'STRICT' , 1 );

Sets a default value override for C<Some::Package> to pick up when it compiles.

Note: This B<MUST> be performed prior to compile-time, or it won't affect the module B<AT ALL>

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Facilitate toggling validation code at users request, without extra performance penalties.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Package-Strictures",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Package::Strictures",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Config::INI::Reader" => "0.014",
    "Import::Into" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "B::Deparse" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "lib" => 0,
    "namespace::clean" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "B::Deparse" => 0,
  "Carp" => 0,
  "Config::INI::Reader" => "0.014",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Import::Into" => 0,
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "constant" => 0,
  "lib" => 0,
  "namespace::clean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
